### PR TITLE
Get list of collateral tokens from Kleros TCR

### LINKED ISSUE #422: Get list of collateral tokens from Kleros TCR

### PR DESCRIPTION
**Pull request by mariano-aguero**
Closes #422 

#### Includes
- Added a service to fetch tokens from kleros interaction contract
- We can validate the amount of tokens, we have 395 in https://tokens.kleros.io/tokens , and the same quantity in the dropdown to select
- We don't have the contract deployed in the docker, also doesn't exist in rinkeby, so if the user use rinkeby or ganache we choose the old tokens that we have in the networks file.

### CODE REVIEW

**Comment by mariano-aguero** [comment]
I refactor the old code to a more uniswap style, we only fetch tokens with a badge. 
So first of all, we fetch addresses of tokens that have the badge. Then we use a[ view contract](https://etherscan.io/address/0xf9b9b5440340123b21bff1ddafe1ad6feb9d6e7f#code) to fetch their submission IDs on the TCR. Later we return all the available token data at once. Warning: some token contracts do not implement the decimals function, I added a fallback to 18 decimals.
To take into account, the TCR contract contains all token submissions, a token submission contains the token's data its current status and request history (name, ticker, address, decimals, etc)

File: app/src/services/kleros.ts [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+      // With the token IDs, get the information and add it to the object.
+      const fetchedTokens = await this.tokensViewContract.getTokens(klerosTCRAddress, submissionIDs)
+      const tokens = fetchedTokens
+        .filter((tokenInfo: any) => tokenInfo[3] !== ethers.constants.AddressZero)
+        .map(
+          (token: any): Token => {
+            return {
+              symbol: token.ticker,
+              address: token.addr,
+              decimals: token.decimals.toNumber() || 18, // Token does not implement the 'decimals()' function, falling back to the default, 18 decimal places.
```
Comment (mariano-aguero) [inline app/src/services/kleros.ts]: I think we need consensus right here, how to manage this situation. Some tokens like DGD may not implement the decimals function, maybe we can manage some situations with an array of constants or another pattern, what do you think ?

File: app/src/util/networks.ts [hunk 2]
```diff
@@ -75,6 +81,9 @@ const networks: { [K in NetworkId]: Network } = {
       marketMakerFactory: '0x89F54625565792729fd3Fa5F689c46e7612b4E8b',
       conditionalTokens: '0xe6Cdc22F99FD9ffdC03647C7fFF5bB753a4eBB21',
       oracle: '0x5A90132C104CA33652Ee71E2E645D5A4d8E5d7D6',
+      klerosBadge: '0x0',
+      klerosTokenView: '0x0',
+      klerosTCR: '0x0',
```
Comment (mariano-aguero) [inline app/src/util/networks.ts]: We don't have support for rinkeby :(

File: app/src/services/kleros.ts [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+import { getContractAddress, getTokensByNetwork, networkIds } from '../util/networks'
+import { Web3Provider } from 'ethers/providers'
+
+const klerosBadgeAbi = [
+  'function queryAddresses(address _cursor, uint _count, bool[8] _filter, bool _oldestFirst) external view returns (address[] values, bool hasMore)',
+]
+
+const klerosTokensViewAbi = [
+  'function getTokensIDsForAddresses(address _t2crAddress, address[] _tokenAddresses ) external view returns (bytes32[] result)',
+  'function getTokens(address _t2crAddress, bytes32[] _tokenIDs ) external view returns (tuple(bytes32 ID, string name, string ticker, address addr, string symbolMultihash, uint8 status, uint256 decimals)[] result)',
```
Comment (mariano-aguero) [inline app/src/services/kleros.ts]: Check this, how to manage a struct (tuple) with human readable abi

File: app/src/services/kleros.ts (hunk 1 above)
Comment (fvictorio) [reply app/src/services/kleros.ts]: Kleros's Uniswap fork falls back to 18, so I think it's ok to do the same.

File: app/src/services/kleros.ts (hunk 3 above)
Comment (fvictorio) [reply app/src/services/kleros.ts]: Well, if it works, it works.

**Comment by fvictorio** [comment]
I'll try to deploy this in Rinkeby and our docker image.

File: app/src/services/kleros.ts [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+        addressesWithBadge = addressesWithBadge.concat(
+          result[0].filter((address: string) => address !== ethers.constants.AddressZero),
+        )
+        hasMore = result.hasMore
+      }
+
+      // Fetch their submission IDs on the TCR.
+      const network = await this.provider.getNetwork()
+      const networkId = network.chainId
+      const klerosTCRAddress = getContractAddress(networkId, 'klerosTCR')
```
Comment (fvictorio) [inline app/src/services/kleros.ts]: I would add this as a constructor parameter and then use it as an instance variable.

File: app/src/services/kleros.ts [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+            }
+          },
+        )
+
+      const basicOrder = ['dai', 'cdai', 'weth', 'usdc', 'owl', 'chai'].reverse()
+
+      return tokens.sort((tokenA: Token, tokenB: Token) => {
+        let sortBy: number
+
+        const indexOfB = basicOrder.indexOf(tokenB.symbol.toLowerCase())
```
Comment (fvictorio) [inline app/src/services/kleros.ts]: This is going to be extremely pedantic but I'm going to do it anyway: A should go before B.

File: app/src/services/kleros.ts [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+  tokensViewContract: Contract
+
+  constructor(
+    badgeContractAddress: string,
+    tokensViewContractAddress: string,
+    provider: Web3Provider,
+    signerAddress: Maybe<string>,
+  ) {
+    this.provider = provider
+    if (signerAddress) {
```
Comment (fvictorio) [inline app/src/services/kleros.ts]: I know this is the way we always do it, but I'm wonder if

```typescript
      this.badgeContract = new ethers.Contract(
        badgeContractAddress,
        klerosBadgeAbi,
        provider,
      )
      this.tokensViewContract = new ethers.Contract(
        tokensViewContractAddress,
        klerosTokensViewAbi,
        provider,
      )
      if (signerAddress) {
        this.badgeContract = this.badgeContract.connect(signer)
        this.tokensViewContract =  this.tokensViewContract(signer)
      }
```

isn't better. I'm on the fence with this one so either approach is ok by me.

File: app/src/services/kleros.ts (hunk 4 above)
Comment (mariano-aguero) [reply app/src/services/kleros.ts]: Done!

File: app/src/services/kleros.ts (hunk 5 above)
Comment (mariano-aguero) [reply app/src/services/kleros.ts]: Done!

File: app/src/services/kleros.ts (hunk 6 above)
Comment (mariano-aguero) [reply app/src/services/kleros.ts]: Done!

**Comment by fvictorio** [comment]
I was able to deploy this to ganache, but without tokens. Adding our tokens to the list it's surely doable but I think it's going to take me a lot of time that I don't think it's worth it. Let's continue with the approach of using our hardcoded tokens in ganache for now. If at some point is really important to have this feature in a network other than mainnet, we can just add Kovan to our supported networks.

I will review the code again when the merge conflicts are resolved.

**Comment by mariano-aguero** [comment]
> I was able to deploy this to ganache, but without tokens. Adding our tokens to the list it's surely doable but I think it's going to take me a lot of time that I don't think it's worth it. Let's continue with the approach of using our hardcoded tokens in ganache for now. If at some point is really important to have this feature in a network other than mainnet, we can just add Kovan to our supported networks.
> 
> I will review the code again when the merge conflicts are resolved.

I solved the conflicts ...

**Pull request merged**